### PR TITLE
Added deduplication headers to rodeos rabbit stream 📦

### PR DESCRIPTION
**Pull request by ndcgundlach**
Added deduplication headers to rodeos rabbit stream

Updated the transactional amqp publisher to optionally add the x-deduplication-header to enable the de-duplication of messages.  Rodeos set the option to true.

## Change Type
**Select *ONE*:**
- [ ] Documentation

- [ ] Stability bug fix

- [ x] Other
performance improvement
- [ ] Other - special case

## Testing Changes
**Select *ANY* that apply:**
- [ ] New Tests

- [ ] Existing Tests

- [ ] Test Framework

- [ ] CI System

- [ ] Other

## Consensus Changes
- [ ] Consensus Changes

## API Changes
- [ ] API Changes

## Documentation Additions
- [x] Documentation Additions

### CODE REVIEW

File: libraries/amqp/transactional_amqp_publisher.cpp [hunk 1]
```diff
@@ -145,6 +150,11 @@ void transactional_amqp_publisher_impl::pump_queue() {
    in_flight = true;
    for(const auto& msg : message_deque) {
       AMQP::Envelope envelope(msg.second.data(), msg.second.size());
+      AMQP::Table headers;
```
Comment (linhuang-blockone) [inline libraries/amqp/transactional_amqp_publisher.cpp]: A minor one. You can move this declaration inside the if statement below. The same is for the next one.

**Review by linhuang-blockone — APPROVED** [review]
How easy to add a test to it?

File: tests/rodeos_rabbit_test.py [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+###############################################################
+# rodeos_test.py
+#
+# rodeos integration test 
+# 
+# This test creates a producer node with state history plugin and a 
+# rodeos process with a test filter to connect to the producer. Pushes 
+# a few transactions to the producer and query rodeos get_block endpoint
+# to see if it sees one of the block containing the pushed transaction. 
+# Lastly, it verifies if rodeos get_info endpoint returns a head_block_num.
```
Comment (heifner) [inline tests/rodeos_rabbit_test.py]: Update this comment.

File: tests/rodeos_rabbit_test.py [hunk 3]
```diff
@@ -0,0 +1,286 @@
...
+        for i in range(1,cfTrxBlockNum+1):
+            response = rodeos.get_block(i)
+            if "block_num" in response:
+                if first_block_num == 0:
+                    first_block_num = i
+                assert response["block_num"] == i, "Rodeos responds with wrong block {0}, response body = {1}".format(i, json.dumps(response))
+        Utils.Print("First block in rodeos is {}".format(first_block_num))
+        killTestLoopCount = 3
+        for n in range(killTestLoopCount):
+            # Test sigkill restart
```
Comment (heifner) [inline tests/rodeos_rabbit_test.py]: Why include sigkill in this test?

File: tests/rodeos_rabbit_test.py [hunk 4]
```diff
@@ -0,0 +1,286 @@
...
+from testUtils import BlockLogAction
+import json
+import sys
+import signal
+import time
+import subprocess
+import os
+import shutil
+
+###############################################################
```
Comment (heifner) [inline tests/rodeos_rabbit_test.py]: I don't see how this tests de-dup of streamed rabbit messages. Please explain.

**Review by heifner — CHANGES_REQUESTED** [review]

File: programs/rodeos/streams/rabbitmq.hpp [hunk 5]
```diff
@@ -41,14 +42,6 @@ class rabbitmq : public stream_handler {
...
    {
       ilog("Connecting to RabbitMQ address ${a} - Queue: ${q}...", ("a", address)( "q", queue_name_));
-      std::atomic<bool> error = false;
-      eosio::amqp_handler declare_queue( address_, fc::seconds(60), fc::milliseconds(200), [&error](const std::string& err){
-         elog("AMQP Queue error: ${e}", ("e", err));
-         appbase::app().quit();
-         error = true;
-      } );
-      if( !error ) declare_queue.declare_queue(queue_name_);
-      if( error ) return;
```
Comment (heifner) [inline programs/rodeos/streams/rabbitmq.hpp:51]: I guess we don't currently have any tests that needed this. In the future any streaming test will need to create the queue manually.

**Review by heifner — APPROVED** [review]

**Pull request merged**